### DISCUSSION TITLE
UserWarning: The setup method 'add_url_rule' can no longer be called on the blueprint 'api'. It has already been registered at least once, any changes will not be applied consistently.

### DISCUSSION THREAD

**Discussion by sebvargo** [pallets/flask#4940] [Q&A]
Hello. I am launching a Flask app and keep getting this warning. I thought adding an app factory function and creating an app context would solve the issue, but the warning persists:

> **UserWarning:** The setup method 'add_url_rule' can no longer be called on the blueprint 'api'. It has already been registered at least once, any changes will not be applied consistently. Make sure all imports, decorators, functions, etc. needed to set up the blueprint are done before registering it.
> This warning will become an exception in Flask 2.3.
>   app_or_blueprint.add_url_rule(self.prefix or "/", "root", self.render_root)

This is my project setup:

|   main.py
+---app
|   |   __init__.py
|   |   
|   +---api
|   |   |   routes.py
|   |   |   __init__.py
|   |   |   
|   |   \---__pycache__
|   |           routes.cpython-39.pyc
|   |           __init__.cpython-39.pyc
|   |           
|   +---apidocs
|   |   |   routes.py
|   |   |   __init__.py
|   |   |   
|   +---templates
|   |   +---api
|   |   \---apidocs
|   |           home.html

**main.py**
```
from app import create_app
from dotenv import load_dotenv
import os

load_dotenv()   # load .env file into environment variables

if __name__ == '__main__':
    app = create_app()
    
    if os.getenv("APP_ENVIRONMENT") == "dev":
        app.debug=True
    else: 
        app.debug=False
    app.run()
```

**app/__init__py**
```
from flask import Flask

def create_app():
    app = Flask(__name__)
    
    with app.app_context():
        from app.api import BP as bp_api
        from app.apidocs import BP as bp_apidocs
        app.register_blueprint(bp_api, url_prefix='/api')
        app.register_blueprint(bp_apidocs, url_prefix='/apidocs')
    
    return app
```
**app/api/__init__.py**
```
from flask import Blueprint
from flask_restx import Api

BP = Blueprint('api', __name__)
API = Api(BP)

from app.api import routes
```

**Comment by tachyondecay** [comment]
Without the full traceback, it’s difficult to see where the warning is coming from.

**Reply by ThiefMaster** [reply]
warnings do not come with full tracebacks ;) but a nice trick to get one anyway is to configure the warning to be an error instead.
